runtime/fastly/host-api/host_api.cpp

```diff
@@ -10,26 +10,92 @@
+  std::vector<fastly_compute_at_edge_async_io_handle_t> handles;
+  handles.reserve(tasks_len);
+  uint64_t now = 0;
+  uint64_t soonest_deadline = 0;
+  size_t soonest_deadline_idx = -1;
+  for (size_t idx = 0; idx < tasks_len; ++idx) {
+    auto task = tasks->at(idx);
+    uint64_t deadline = task->deadline();
+    // Select for completed task deadlines before performing the task select host call.
+    if (deadline > 0) {
+      MOZ_ASSERT(task->id() == NEVER_HANDLE);
+      if (now == 0) {
+        now = MonotonicClock::now();
+        MOZ_ASSERT(now > 0);
+      }
+      if (deadline <= now) {
+        return idx;
+      }
+      if (soonest_deadline == 0 || deadline < soonest_deadline) {
+        soonest_deadline = deadline;
+        soonest_deadline_idx = idx;
+      }
+    } else {
+      uint32_t handle = task->id();
+      // Timer task handles are skipped and never passed to the host.
+      MOZ_ASSERT(handle != NEVER_HANDLE);
+      handles.push_back(handle);
+    }
+  }
+
+  // When there are no async tasks, sleep until the deadline
+  if (handles.size() == 0) {
+    MOZ_ASSERT(soonest_deadline > 0);
+    while (soonest_deadline > now) {
+      uint64_t duration = soonest_deadline - now;
+      timespec req{.tv_sec = static_cast<time_t>(duration / 1000000000),
+                   .tv_nsec = static_cast<long>(duration % 1000000000)};
+      timespec rem;
+      nanosleep(&req, &rem);
+      now = MonotonicClock::now();
+    }
+    return soonest_deadline_idx;
   }
-  fastly_world_list_handle_t hs{.ptr = handles, .len = tasks_len};
+
+  fastly_world_list_handle_t hs{.ptr = handles.data(), .len = handles.size()};
   fastly_world_option_u32_t ret;
   fastly_compute_at_edge_types_error_t err = 0;
-  if (!fastly_compute_at_edge_async_io_select(&hs, 0, &ret, &err)) {
+  if (!fastly_compute_at_edge_async_io_select(&hs, (soonest_deadline - now) / 1000000, &ret,
+                                              &err)) {
     abort();
   } else if (ret.is_some) {
-    return ret.val;
-  } else {
+    // The host index will be the index in the list of tasks with the timer tasks filtered out.
+    // We thus need to offset the host index by any timer tasks appearing before the nth
+    // non-timer task.
+    size_t task_idx = 0;
+    for (size_t idx = 0; idx < tasks_len; ++idx) {
+      if (tasks->at(idx)->id() != NEVER_HANDLE) {
+        if (ret.val == task_idx) {
+          return idx;
+        }
+        task_idx++;
+      }
+    }
     abort();
+  } else {
+    // No value case means a timeout, which means soonest_deadline_idx is set.
+    MOZ_ASSERT(soonest_deadline > 0);
+    MOZ_ASSERT(soonest_deadline_idx != -1);
+    // Verify that the task definitely is definitely ready from a time perspective, and if not
+    // repeat the entire routine as we cannot progress a timer before it's actually ready.
+    now = MonotonicClock::now();
+    if (soonest_deadline > now) {
+      return api::AsyncTask::select(tasks);
```

Comment:
It might be worth reworking this to have a top-level loop instead of a recursive call. There could be a situation where the stars align and the stack is blown.

Reply:
I've added an extra sleep here to correct it.